pkg/cluster/kubeclient/roundtrip.go

```diff
@@ -0,0 +1,166 @@
+package kubeclient
+
+import (
+	"errors"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd/api/v1"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+	"github.com/openshift/openshift-azure/pkg/util/managedcluster"
+)
+
+// The retryingRoundTripper implementation is customised to help with multiple
+// network connection-related issues seen in CI which we haven't necessarily
+// been able to fully explain yet.  Importantly, it is not yet clear whether any
+// of these issues could impact cluster end-users or whether they are more
+// symptomatic of a CI-related issue.
+//
+// 1. We think the following flow may be possible:
+//    * client has an open TCP connection to master-000000:443 via LB.
+//    * master-000000 is deallocated as part of a rotation.
+//    * deallocation takes place too quickly for TCP connection to be closed
+//      properly.
+//    * next client request errors.
+//    We're trying to solve this via the disableKeepAlives flag: in such
+//    circumstances the caller can set this and a new TCP connection will be
+//    opened for each request.
+//
+// 2. Notwithstanding the fix in 1, we are still seeing read: connection timed
+//    out errors, such as "WaitForInfraDaemonSets: Get
+//    https://hostname.eastus.cloudapp.azure.com/apis/apps/v1/namespaces/default/daemonsets/router:
+//    read tcp 172.16.217.87:48274->52.188.220.9:443: read: connection timed
+//    out".  At least most of these errors are on GETs, and it appears that the
+//    timeout in question is long (around 15-16 minutes).  Current best guess is
+//    we're hitting the kernel tcp_retries2 limit; it looks like the client
```

Comment:
Little bit offtopic theory question: Should client send one more package in this case if it didn't receive ACK for initial package?

Reply:
yes, afaics it does this with exponential backoff for 15 minutes (kernel configurable) before giving up.